### PR TITLE
Merge caffe2_specific into ROCm/master

### PR DESCRIPTION
**Pull request by rohithkrn**
@iotamudelta @petrex

### CODE REVIEW

**Comment by iotamudelta** [comment]
@pytorchbot retest this please

@rohithkrn should we squash and merge this as per Mayank's talk on Friday or do you anticipate that you will need to cherrypick parts of this PR for upstreaming?

**Comment by iotamudelta** [comment]
@rohithkrn looks like legit CI testing issues.

**Comment by rohithkrn** [comment]
@iotamudelta will look into the failures. Once it passes, we can squash. I don't intend to cherry-pick.

**Comment by ashishfarmer** [comment]
The only failing test is conv gradients with group=2 using MIOpen. It is a regression, will have the fix in soon

File: caffe2/python/core.py [hunk 1]
```diff
@@ -2110,12 +2112,13 @@ def DeduplicateGradientSlices(self, g, aggregator='sum'):
...
         device_option = caffe2_pb2.DeviceOption()
-        device_option.device_type = caffe2_pb2.CUDA
+        device_option.device_type = caffe2_pb2.HIP if workspace.has_hip_support else caffe2_pb2.CUDA
         device_option.cuda_gpu_id = gpu_id
+        device_option.hip_gpu_id = gpu_id
         self._net.device_option.CopyFrom(device_option)
         if use_cudnn:
             for op in self._net.op:
-                op.engine = "CUDNN"
+                op.engine = 'CUDNN'
```
Comment (iotamudelta) [inline caffe2/python/core.py]: Gratuitous?

File: caffe2/python/convnet_benchmarks.py [hunk 2]
```diff
@@ -723,5 +803,8 @@ def GetArgumentParser():
             'VGGA': VGGA,
             'Inception': Inception,
             'MLP': MLP,
+            'Resnet50': Resnet50,
+            'Inception_v2':Inception_v2
+
```
Comment (iotamudelta) [inline caffe2/python/convnet_benchmarks.py]: Gratuitous white space change

File: .jenkins/caffe2/amd/install_dependencies_amd.sh [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+git clone --recursive https://github.com/ROCmSoftwarePlatform/Thrust.git /data/Thrust
+rm -rf /data/Thrust/thrust/system/cuda/detail/cub-hip
+git clone --recursive https://github.com/ROCmSoftwarePlatform/cub-hip.git /data/Thrust/thrust/system/cuda/detail/cub-hip
+
+
+export PATH=/opt/rocm/bin:$PATH
+export PATH=/opt/rocm/hcc/bin:$PATH
+export PATH=/opt/rocm/hip/bin:$PATH
+export PATH=/opt/rocm/opencl/bin:$PATH
+export THRUST_ROOT=/data/Thrust
```
Comment (iotamudelta) [inline .jenkins/caffe2/amd/install_dependencies_amd.sh]: Note: this will hopefully change soon! :-)

**Review by iotamudelta — COMMENTED** [review]
A few things I noticed. Given our experience, upstream didn't like too many HIP-specific changes in the tests. We shall see, I guess.

File: caffe2/python/workspace_test.py [hunk 4]
```diff
@@ -317,7 +317,7 @@ def testCreateWorkspace(self):
         self.assertTrue("test" in workspaces)
 
 
-@unittest.skipIf(not workspace.has_gpu_support, "No gpu support.")
+@unittest.skipIf(not workspace.has_gpu_support and not workspace.has_hip_support, "No gpu support.")
```
[the remark this answers is not in the document]
Comment (rohithkrn) [reply caffe2/python/workspace_test.py]: This is skip if no cuda and no hip

[the remark this answers is not in the document]
Comment (iotamudelta) [reply caffe2/python/workspace_test.py]: I meant in general, but sure. :-)

File: caffe2/python/workspace_test.py [hunk 5]
```diff
@@ -339,13 +339,22 @@ def testFetchBlobGPU(self):
         self.assertEqual(fetched_again.shape, (1, 2, 3, 4))
         np.testing.assert_array_equal(fetched_again, 2.0)
 
+    @unittest.skipIf(not workspace.has_gpu_support, "No gpu support.")
```
[the remark this answers is not in the document]
Comment (rohithkrn) [reply caffe2/python/workspace_test.py]: No, its gpu. It's to make sure cuda test not run on hip device.

File: caffe2/python/optimizer.py [hunk 6]
```diff
@@ -1453,8 +1453,7 @@ def build_ftrl(model, engine="SIMD", **kwargs):
 
 
 def build_gftrl(model, engine="", **kwargs):
-    if engine == "SIMD":
-        assert core.IsOperator('GFtrl_ENGINE_SIMD')
+    # SIMD version of GFTRL is not supported
```
[the remark this answers is not in the document]
Comment (rohithkrn) [reply caffe2/python/optimizer.py]: Looks like a mis-merge

File: caffe2/python/operator_test/activation_ops_test.py [hunk 7]
```diff
@@ -16,9 +16,9 @@
 import unittest
 
 
-class TestActivations(serial.SerializedTestCase):
+class TestActivations(hu.HypothesisTestCase):
```
[the remark this answers is not in the document]
Comment (rohithkrn) [reply caffe2/python/operator_test/activation_ops_test.py]: Thanks for pointing. It's an error on my end.

**Comment by rohithkrn** [comment]
@ashishfarmer could you please review the changes in resnet benchmarking.

File: caffe2/python/convnet_benchmarks.py [hunk 8]
```diff
@@ -563,45 +565,81 @@ def Inception(order, cudnn_ws, model_path=""):
...
-        init_def.ParseFromString(f.read())
-        init_def.device_option.CopyFrom(device_opts)
+def Resnet50(args):
+    gpus = [0]
+    device_opt = core.DeviceOption(caffe2_pb2.HIP)
+    device_opt.hip_gpu_id = gpus[0]
+
+    # Batch size of 32 sums up to roughly 5GB of memory per device
+    batch_per_device = args.batch_size
+    total_batch_size = batch_per_device * len(gpus)
```
Comment (ashishfarmer) [inline caffe2/python/convnet_benchmarks.py]: This is not necessary. It is a single GPU test. batch size and total batch size will always be the same

File: caffe2/python/convnet_benchmarks.py [hunk 9]
```diff
@@ -563,45 +565,81 @@ def Inception(order, cudnn_ws, model_path=""):
...
+
+    num_labels = 1000
+
+    base_learning_rate = 0.0004 * total_batch_size
+
+    # Weight decay (L2 regularization)
+    weight_decay = 1e-4
+
+    num_epochs = 10
```
Comment (ashishfarmer) [inline caffe2/python/convnet_benchmarks.py]: Unused variable

File: caffe2/python/convnet_benchmarks.py [hunk 10]
```diff
@@ -563,45 +565,81 @@ def Inception(order, cudnn_ws, model_path=""):
...
-    device_opts.device_type = caffe2_pb2.HIP
-    device_opts.hip_gpu_id = 0
 
-    INIT_NET_PB = os.path.join(model_path, "init_net.pb")
-    PREDICT_NET_PB = os.path.join(model_path, "predict_net.pb")
-    init_def = caffe2_pb2.NetDef()
-    with open(INIT_NET_PB, 'rb') as f:
-        init_def.ParseFromString(f.read())
-        init_def.device_option.CopyFrom(device_opts)
+def Resnet50(args):
```
Comment (ashishfarmer) [inline caffe2/python/convnet_benchmarks.py]: Could just use the signature Resnet50(order, cudnn_ws, model_path="") just like the other model creator functions

File: caffe2/python/convnet_benchmarks.py [hunk 11]
```diff
@@ -654,7 +692,10 @@ def AddParameterUpdate(model):
 
 
 def Benchmark(model_gen, arg):
-    model, input_size = model_gen(arg.order, arg.cudnn_ws, arg.model_path)
+    if arg.model == 'Resnet50':
```
Comment (ashishfarmer) [inline caffe2/python/convnet_benchmarks.py]: You won't need this once you change the method signature to Resnet50(order, cudnn_ws, model_path="")

File: caffe2/python/convnet_benchmarks.py (hunk 9 above)
Comment (rohithkrn) [reply caffe2/python/convnet_benchmarks.py]: Will fix.

File: caffe2/python/convnet_benchmarks.py (hunk 10 above)
Comment (rohithkrn) [reply caffe2/python/convnet_benchmarks.py]: Resnet needs other args as well, like args.forward_only, args.batch_size which necessitates a different signature.

**Review by ashishfarmer — APPROVED** [review]

File: caffe2/python/convnet_benchmarks.py [hunk 12]
```diff
@@ -723,5 +837,8 @@ def GetArgumentParser():
             'VGGA': VGGA,
             'Inception': Inception,
             'MLP': MLP,
+            'Resnet50': Resnet50,
+            'Inception_v2':Inception_v2
         }
         Benchmark(model_map[args.model], args)
+
```
Comment (iotamudelta) [inline caffe2/python/convnet_benchmarks.py]: Gratuitous white space change

File: caffe2/python/optimizer.py [hunk 13]
```diff
@@ -1454,7 +1456,7 @@ def build_ftrl(model, engine="SIMD", **kwargs):
 
 def build_gftrl(model, engine="", **kwargs):
     if engine == "SIMD":
-        assert core.IsOperator('GFtrl_ENGINE_SIMD')
+        assert core.IsOperator('GFtrl_ENGINE_SIMD')
```
Comment (iotamudelta) [inline caffe2/python/optimizer.py]: Gratuitous white space change.

File: caffe2/python/workspace.py [hunk 14]
```diff
@@ -47,7 +47,7 @@
     NumCudaDevices = C.num_cuda_devices
     GetCUDAVersion = C.get_cuda_version
     GetCuDNNVersion = C.get_cudnn_version
-
+    NumGpuDevices = NumCudaDevices
```
Comment (iotamudelta) [inline caffe2/python/workspace.py]: Missing white space, I'd assume this would fail flake.

**Review by iotamudelta — COMMENTED** [review]
Some more white space comments. Sorry, but they tend to come up when upstreaming.

File: caffe2/python/operator_test/group_conv_test.py [hunk 15]
```diff
@@ -65,8 +66,8 @@ def test_group_convolution(
             w = w.transpose((0, 3, 1, 2))
 
         inputs = [X, w, b] if use_bias else [X, w]
-
-        self.assertDeviceChecks(dc, op, inputs, [0])
+        if order != 'NHWC' or group == 1:
```
Comment (ashishfarmer) [inline caffe2/python/operator_test/group_conv_test.py]: Could we re-check this

File: rocm-docs/caffe2-build.md [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+git clone --recurse-submodules -b caffe2_specific git@github.com:ROCmSoftwarePlatform/pytorch.git
+```
+Navigate to repo directory
+```
+cd pytorch
+```
+
+### Launch the docker container
+```	
+docker run -it --network=host --device=/dev/kfd --device=/dev/dri --group-add video -v $PWD:/pytorch rocm/caffe2:unbuilt-rocm1.8.2-v3
```
Comment (ashishfarmer) [inline rocm-docs/caffe2-build.md]: Correct docker tag

File: rocm-docs/caffe2-build.md [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+Caffe2 benchmarking script supports the following networks.
+1. MLP
+2. AlexNet
+3. OverFeat
+4. VGGA
+5. Inception
+6. Inception_v2
+7. Resnet50 
+
+*Special case:* Inception_v2 and Resnet50 will need their corresponding protobuf files to run the benchmarks. Protobufs can be downloaded from caffe2 model zoo using the below command. Substitute model_name with `inception_v2` or `resnet50`
```
Comment (ashishfarmer) [inline rocm-docs/caffe2-build.md]: No need for resnet50 protobuf download anymore

**Review by ashishfarmer — CHANGES_REQUESTED** [review]

File: caffe2/python/operator_test/group_conv_test.py (hunk 15 above)
Comment (rohithkrn) [reply caffe2/python/operator_test/group_conv_test.py]: I feel it's required.

**Review by ashishfarmer — APPROVED** [review]
